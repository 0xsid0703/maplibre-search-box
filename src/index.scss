.stadiamaps-search-box {
  $spinner-height: 20px;
  width: 280px;

  &.loading {
    .spinner {
      display: block !important;
    }

    .cancel {
      display: none !important;
    }
  }

  .input-container {
    input {
      border-radius: 5px;
      font-size: 1rem;
      height: 24px;
      padding: 5px 25px 5px 5px;
      width: 250px;
    }

    .spinner {
      position: absolute;
      right: 5px;
      top: 8px;
    }

    .spinner.hidden {
      display: none;
    }

    .cancel {
      background: #aaa;
      border-radius: 50%;
      color: #fff;
      cursor: pointer;
      display: block;
      font-size: 1.5rem;
      height: 20px;
      position: absolute;
      right: 5px;
      top: 9px;
      width: 20px;

      &::before {
        content: "\02a2f";
        display: block;
        margin-left: 4.5px;
        margin-top: -4.5px;
      }
    }

    .cancel.hidden {
      display: none;
    }
  }

  .results {
    background: #fefefe;
    border-radius: 5px;
    margin: 10px 0;
    width: 284px;

    &.hidden {
      display: none;
    }

    .results-list {
      max-height: 500px;
      overflow-y: scroll;

      &:first-child {
        border-top-left-radius: 5px;
        border-top-right-radius: 5px;
      }
    }

    .result {
      cursor: pointer;
      font-size: 1rem;
      padding: 5px;

      &:hover {
        background: #dfdfdf;
      }
    }

    .result-label {
      font-weight: 500;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .result-extra {
      font-size: 0.7rem;
      font-style: italic;
      text-transform: uppercase;
    }
  }

  .search-attribution {
    padding: 5px;

    a {
      &:visited {
        color: #000;
      }
    }
  }

  .lds-dual-ring {
    display: inline-block;
    height: $spinner-height;
    width: $spinner-height;
  }

  .lds-dual-ring::after {
    animation: lds-dual-ring 1.2s linear infinite;
    border: $spinner-height * 0.075 solid #fff;
    border-color: #ccc transparent;
    border-radius: 50%;
    content: " ";
    display: block;
    height: $spinner-height * 0.8;
    margin: $spinner-height * 0.1;
    width: $spinner-height * 0.8;
  }

  @keyframes lds-dual-ring {
    0% {
      transform: rotate(0deg);
    }

    100% {
      transform: rotate(360deg);
    }
  }
}
