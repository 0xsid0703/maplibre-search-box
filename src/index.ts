import type { IControl, Map } from "maplibre-gl";
import {
  type AutocompleteRequest,
  GeocodingApi,
  PeliasGeoJSONFeature,
  PeliasLayer,
} from "@stadiamaps/api";
import "./index.scss";
import logo from "./logo.svg";

export class MapLibreSearchControlOptions {
  useMapFocusPoint = true;
  mapFocusPointMinZoom = 5;
  fixedFocusPoint: [number, number] = null;
  searchOnEnter = false;
  maxResults = 5;
  minInputLength = 3;
  minWaitPeriodMs = 100;
  layers: PeliasLayer[] = null;
}

export class MapLibreSearchControl implements IControl {
  private map: Map | null = null;
  private container: HTMLElement | null = null;
  private resultsContainer: HTMLElement | null = null;
  private resultsList: HTMLElement | null = null;
  private input: HTMLInputElement | null = null;
  private clearButton: HTMLElement | null = null;
  private loadingSpinner: HTMLElement | null = null;
  private api = new GeocodingApi();
  private lastRequestAt = 0;
  private lastRequestString = "";
  private resultFeatures: PeliasGeoJSONFeature[] = [];

  options = new MapLibreSearchControlOptions();

  constructor(options: object | MapLibreSearchControlOptions) {
    if (options !== undefined) {
      this.options = {
        ...this.options,
        ...options,
      };
    }
  }

  onAdd(map: Map): HTMLElement {
    this.map = map;
    this.container = this.buildInput();
    return this.container;
  }

  buildInput(): HTMLElement {
    const container = document.createElement("div");
    container.className = "maplibregl-ctrl stadiamaps-search-box";
    const inputContainer = container.appendChild(document.createElement("div"));
    inputContainer.className = "input-container";

    this.clearButton = inputContainer.appendChild(
      document.createElement("span")
    );
    this.clearButton.className = "cancel hidden";
    this.clearButton.addEventListener("click", this.onClear.bind(this));

    this.loadingSpinner = inputContainer.appendChild(
      document.createElement("span")
    );
    this.loadingSpinner.className = "spinner lds-dual-ring hidden";

    this.input = inputContainer.appendChild(document.createElement("input"));
    this.input.type = "text";
    this.input.placeholder = "Search for places...";
    this.input.addEventListener("input", this.onInput.bind(this));
    this.input.addEventListener("focus", this.onFocus.bind(this));
    this.input.addEventListener("keypress", this.onKey.bind(this));
    this.input.enterKeyHint = "search";

    this.resultsContainer = container.appendChild(
      document.createElement("div")
    );
    this.resultsContainer.className = "results hidden";
    this.resultsList = this.resultsContainer.appendChild(
      document.createElement("div")
    );
    this.resultsList.className = "results-list";

    const attribution = this.resultsContainer.appendChild(
      document.createElement("div")
    );
    attribution.className = "search-attribution";
    attribution.innerHTML = `<img height="50" width="50" src="${logo}" alt="Stadia Maps" class="logo"> Powered by <a href="https://stadiamaps.com/" target="_blank">Stadia Maps</a><br>&copy; <a href="https://www.openstreetmap.org/copyright" target="_blank" rel="nofollow">OpenStreetMap</a> contributors &amp; <a href="https://stadiamaps.com/attribution/" target="_blank">others</a>`;

    return container;
  }

  getContainer() {
    return this.container;
  }

  async onFocus(_e: Event) {
    if (this.resultFeatures.length > 0) {
      this.showResults();
    }

    this.maybeShowClearButton();
  }

  maybeShowClearButton() {
    const show = this.input.value.length > 0;

    this.clearButton.classList.toggle("hidden", !show);
  }

  async onKey(e: KeyboardEvent) {
    switch (e.key) {
      case "Enter":
        if (this.options.searchOnEnter) {
          await this.onInput(e, true);
        }
        break;
      case "Escape":
        this.onClear();
        break;
      case "ArrowDown":
      case "ArrowUp":
        // TODO
        break;
    }
  }

  async onInput(e: Event, useSearch = false) {
    this.maybeShowClearButton();

    const searchString = this.input.value;

    if (!searchString) {
      this.clearResults();
      this.hideResults();
      return;
    }

    const timeSinceLastRequest = Date.now().valueOf() - this.lastRequestAt;
    if (
      this.lastRequestAt === 0 ||
      timeSinceLastRequest >= this.options.minWaitPeriodMs
    ) {
      if (searchString.length >= this.options.minInputLength) {
        if (searchString !== this.lastRequestString || useSearch) {
          const params: AutocompleteRequest = {
            text: searchString,
            size: this.options.maxResults,
          };

          if (this.options.layers) {
            params.layers = this.options.layers;
          }

          let focusPoint = this.options.fixedFocusPoint;
          if (this.options.useMapFocusPoint) {
            if (this.map.getZoom() >= this.options.mapFocusPointMinZoom) {
              focusPoint = this.map.getCenter().toArray();
            }
          }

          if (focusPoint) {
            params.focusPointLon = focusPoint[0];
            params.focusPointLat = focusPoint[1];
          }

          if (this.options.layers) {
            params.layers = this.options.layers;
          }

          this.lastRequestString = searchString;
          const requestAt = Date.now().valueOf();
          this.lastRequestAt = requestAt;
          this.container.classList.toggle("loading", true);
          try {
            let results;
            if (useSearch) {
              results = await this.api.search(params);
            } else {
              results = await this.api.autocomplete(params);
            }

            // Make sure we only use the latest request
            if (this.lastRequestAt === requestAt) {
              this.clearResults();
              this.resultFeatures = results.features;
              if (this.resultFeatures.length > 0) {
                for (const result of this.resultFeatures) {
                  this.addResult(result);
                }
              } else {
                this.onNoResults();
              }
            }
          } catch (e) {
            console.warn("Something when wrong with the request.");
          } finally {
            if (this.lastRequestAt === requestAt) {
              this.container.classList.toggle("loading", false);
            }
          }
        }
      }
    } else {
      setTimeout(() => {
        this.onInput(e, useSearch);
      }, this.options.minWaitPeriodMs);
    }
  }

  onNoResults() {
    this.showResults();

    const el = document.createElement("div");
    el.className = "result no-result";
    el.textContent = "No Results Found";

    this.resultsList.appendChild(el);
  }

  onClick(feature: PeliasGeoJSONFeature) {
    if (feature.bbox !== undefined) {
      this.map.fitBounds([
        [feature.bbox[0], feature.bbox[1]],
        [feature.bbox[2], feature.bbox[3]],
      ]);
    } else {
      let zoomTarget;
      switch (feature.properties.layer) {
        case "venue":
        case "address":
          zoomTarget = 15;
          break;
        case "marinearea":
        case "locality":
          zoomTarget = 12;
          break;
        case "empire":
        case "macrocounty":
        case "country":
          zoomTarget = 4;
          break;
        case "dependency":
        case "county":
          zoomTarget = 10;
          break;
        case "continent":
        case "ocean":
          zoomTarget = 2;
          break;
        case "macroregion":
        case "region":
          zoomTarget = 7;
          break;
        case "borough":
        case "localadmin":
          zoomTarget = 12;
          break;
        case "macrohood":
        case "postalcode":
        case "neighbourhood":
        case "street":
          zoomTarget = 14;
          break;
        default:
          zoomTarget = 10;
      }
      this.map.jumpTo({
        center: [
          feature.geometry.coordinates[0],
          feature.geometry.coordinates[1],
        ],
        zoom: zoomTarget,
      });
    }

    this.hideResults();
  }

  onClear() {
    this.input.value = "";
    this.maybeShowClearButton();
    this.clearResults();
    this.hideResults();
    this.input.focus();
  }

  buildResult(result: PeliasGeoJSONFeature): HTMLDivElement {
    const el = document.createElement("div");
    el.className = "result";
    el.onclick = this.onClick.bind(this, result);

    el.title = result.properties.label;

    const label = el.appendChild(document.createElement("div"));
    label.textContent = result.properties.label;
    label.className = "result-label";

    const additionalText = el.appendChild(document.createElement("div"));
    additionalText.className = "result-extra";
    let content = result.properties.layer;
    if (
      result.properties.layer == PeliasLayer.Country &&
      result.properties.continent
    ) {
      content += ` in ${result.properties.continent}`;
    } else if (
      result.properties.layer != PeliasLayer.Continent &&
      result.properties.country
    ) {
      content += ` in ${result.properties.country}`;
    }
    additionalText.textContent = content;

    return el;
  }

  addResult(result: PeliasGeoJSONFeature) {
    this.showResults();
    this.resultsList.appendChild(this.buildResult(result));
  }

  hideResults() {
    this.resultsContainer.classList.toggle("hidden", true);
  }

  showResults() {
    this.resultsContainer.classList.toggle("hidden", false);
  }

  clearResults() {
    this.resultFeatures = [];
    this.resultsList.replaceChildren("");
  }

  onRemove(map: Map): void {
    if (map._container) {
      this.container?.parentNode?.removeChild(this.container);
    }
    this.container = null;
    this.map = null;
  }
}
