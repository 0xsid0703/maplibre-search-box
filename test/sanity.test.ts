import { describe, expect, expectTypeOf, it, vi } from "vitest";
import { MapLibreSearchControl } from "../src";
import { IControl, Map } from "maplibre-gl";

describe("search-control", () => {
  function setupMap() {
    const mapEl = document.createElement("div");
    mapEl.id = "map";
    document.body.appendChild(mapEl);

    const control = new MapLibreSearchControl({});
    const map = new Map({
      container: "map",
      style: "https://tiles.stadiamaps.com/styles/alidade_smooth.json",
      center: [-74.5, 40],
      zoom: 2,
    });
    map.addControl(control);

    return {
      mapEl,
      map,
      control,
    };
  }

  it("exists", () => {
    const control = new MapLibreSearchControl({});

    expect(control).toBeInstanceOf(MapLibreSearchControl);
    expectTypeOf(control).toMatchTypeOf<IControl>();
  });

  it("can be added to map", () => {
    const { mapEl } = setupMap();
    expect(mapEl).toMatchSnapshot();
  });
});
