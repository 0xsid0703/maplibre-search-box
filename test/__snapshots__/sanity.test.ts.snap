// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`search-control > can be added to map 1`] = `
<div
  class="maplibregl-map"
  id="map"
>
  <div
    class="maplibregl-canvas-container maplibregl-interactive maplibregl-touch-drag-pan maplibregl-touch-zoom-rotate"
  >
    <canvas
      aria-label="Map"
      class="maplibregl-canvas"
      height="316"
      role="region"
      style="width: 1200px; height: 158px;"
      tabindex="0"
      width="2400"
    />
  </div>
  <div
    class="maplibregl-control-container"
  >
    <div
      class="maplibregl-ctrl-top-left "
    />
    <div
      class="maplibregl-ctrl-top-right "
    >
      <div
        class="maplibregl-ctrl stadiamaps-search-box"
      >
        <div
          class="input-container"
        >
          <span
            class="cancel hidden"
          />
          <span
            class="spinner lds-dual-ring hidden"
          />
          <input
            placeholder="Search for places and addresses..."
            type="text"
          />
        </div>
        <div
          class="results hidden"
        >
          <div
            class="results-list"
          />
          <div
            class="search-attribution"
          >
            Powered by 
            <a
              href="https://stadiamaps.com/"
              target="_blank"
            >
              Stadia Maps
            </a>
            <br />
            Data Â© 
            <a
              href="https://www.openstreetmap.org/copyright"
              rel="nofollow"
              target="_blank"
            >
              OpenStreetMap
            </a>
             contributors & others
          </div>
        </div>
      </div>
    </div>
    <div
      class="maplibregl-ctrl-bottom-left "
    />
    <div
      class="maplibregl-ctrl-bottom-right "
    >
      <details
        class="maplibregl-ctrl maplibregl-ctrl-attrib maplibregl-attrib-empty"
        open=""
      >
        <summary
          aria-label="Toggle attribution"
          class="maplibregl-ctrl-attrib-button"
          title="Toggle attribution"
        />
        <div
          class="maplibregl-ctrl-attrib-inner"
        />
      </details>
    </div>
  </div>
</div>
`;
